src/rtutils.c

```diff
@@ -641,6 +641,19 @@ static size_t jl_static_show_x_sym_escaped(JL_STREAM *out, jl_sym_t *name) JL_NO
     return n;
 }
 
+// `jl_static_show()` cannot call `jl_subtype()`, for the GC reasons
+// explained in the comment on `jl_static_show_x_()`, below.
+// This function checks if `vt <: Function` without triggering GC.
+static int jl_static_is_function_(jl_datatype_t *vt) {
+    while (vt != jl_any_type) {
```

Comment:
Is this sufficient stopping criteria to make sure we never infinite loop? Should I also check for NULL, for example?

Reply:
Yes, this is fine.

Reply:
thanks! :)

Reply:
because of where this gets called from, yeah, you should probably check for NULL also

Reply:
Yeah, I was just about to come back and say that. This function is regularly called on invalid data. You might also want to include a max trip count or something in case the datatype loops.

Reply:
It shouldn't be quite that invalid, but we might end up here to print out a message about the incomplete initialization of some struct type

Reply:
I usually like to be as defensive as possible in these functions, since I not-irregularly call them using `jl_` on complete garbage data, which could easily have cycles.

Reply:
What is a reasonable max number of iterations? 100? Do we have type hierarchies bigger than that? 1000?

Pointer chasing is pretty fast, so i don't seem much harm in a reasonably large number?

Reply:
Does it make sense to throw an exception if I encounter a null and/or infinite loop? Or should I just print a warning and then return false?

Reply:
Okay, I've pushed up 2479af7, which fixes this, by throwing an error after 10,000 iterations. It happens imperceptibly fast, so it seems fine to me:
```julia
julia> v = ()->2+2
#1 (generic function with 1 method)

julia> vt = typeof(v)
var"#1#2"

julia> vt.super = vt;

julia> x = v;

julia> p = Pipe()
Pipe(RawFD(4294967295) init => RawFD(4294967295) init, 0 bytes waiting)

julia> Base.link_pipe!(p, reader_supports_async=true, writer_supports_async=true)
Pipe(RawFD(21) open => RawFD(20) open, 0 bytes waiting)

julia> ccall(:jl_static_show, Cvoid, (Ptr{Cvoid}, Any), p.in, x)
ERROR: static_show: Exit after 10,000 iterations of supertype(). Presuming invalid cycle in datatype object.
Stacktrace:
 [1] top-level scope
   @ ./REPL[7]:1

julia> @time try; ccall(:jl_static_show, Cvoid, (Ptr{Cvoid}, Any), p.in, x) catch; end
  0.000047 seconds (3 allocations: 160 bytes)
```

---------

I think this should be ready for another look. Thanks both! :)

Reply:
Errors are handled differently in this function and in particular you shouldn't call `jl_error` since it allocates.
Just print a warning is fine.

Reply:
Oh, shoot make sense.

Okay, i'll switch to print a warning. Thanks @yuyichao!

Sorry, one more question:
- `jl_printf((JL_STREAM*)STDERR_FILENO,` or `jl_printf(JL_STDERR,`?
    - I see both examples in this file. I've gone with the first one, since it seems to be used in the `jl_static_*` functions?

Done (for now) in 7e2d96736a:
```julia
julia> ccall(:jl_static_show, Cvoid, (Ptr{Cvoid}, Any), p.in, x)
ERROR: static_show: Exit after 10,000 iterations of supertype(). Presuming invalid cycle in datatype object.
```

Reply:
FWIW, the rest of this code also used to use an iteration maximum, then print `•`, but it was removed in favor of explicit tracking with `struct recur_list *depth`. The helper functions, such as this one, `first_arg_datatype` and `jl_static_show_func_sig` have previously ignored those invalid possibilities.

Reply:
Hrmm those are good points...

@vtjnash - without the check for infinite loops, the above example _definitely does_ infinite loop:
```julia
julia> ccall(:jl_static_show, Cvoid, (Ptr{Cvoid}, Any), p.in, x)
^C^C^C^C^CWARNING: Force throwing a SIGINT
ERROR: ^C^C^C^C^C^CInterruptException:
Stacktrace:
 [1] top-level scope
   @ ./REPL[7]:1
```

Is there some way that I should use this `recur_list *depth`? I'm not exactly sure what you're recommending based on the above insight.

Based on the feedback from above, my current thinking is that I'll just remove the warnings, and return false from this function if we encounter a loop or a NULL, since both of those mean: it's not a function!

Does that sound good to you all?

Reply:
I've pushed up aa77e3528e which makes the above change. Please take another look if you can! :)